File: apps/mobile/lib/features/customers/repositories/customer_repository.dart
import 'package:dio/dio.dart';
import '../../../core/supabase/api_client.dart';
import '../models/customer.dart';

class PaginatedCustomers {
  final List<Customer> customers;
  final int total;
  final int page;
  final int limit;
  final int totalPages;
  final bool hasNext;
  final bool hasPrev;

  PaginatedCustomers({
    required this.customers,
    required this.total,
    required this.page,
    required this.limit,
    required this.totalPages,
    required this.hasNext,
    required this.hasPrev,
  });
}

/// Repository layer for Customers.
/// All HTTP calls go through here — providers never touch Dio directly.
class CustomerRepository {
  final Dio _client = apiClient;

  /// Fetch all customers from the Next.js API with pagination.
  Future<PaginatedCustomers> getCustomers({
    int page = 1,
    int limit = 20,
    String? query,
    String? phone,
    String sortBy = 'created_at',
    String sortOrder = 'desc',
  }) async {
    final Map<String, dynamic> queryParams = {
      'page': page,
      'limit': limit,
      'sort_by': sortBy,
      'sort_order': sortOrder,
    };
    if (query != null && query.isNotEmpty) {
      queryParams['query'] = query;
    }
    if (phone != null && phone.isNotEmpty) {
      queryParams['phone'] = phone;
    }

    final response = await _client.get('/customers', queryParameters: queryParams);

    if (response.statusCode == 200) {
      final data = response.data;
      if (data['success'] == true && data['data'] != null) {
        final customersData = data['data']['customers'] as List;
        return PaginatedCustomers(
          customers: customersData.map((e) => Customer.fromJson(e)).toList(),
          total: data['data']['total'] ?? 0,
          page: data['data']['page'] ?? 1,
          limit: data['data']['limit'] ?? 20,
          totalPages: data['data']['total_pages'] ?? 1,
          hasNext: data['data']['has_next'] ?? false,
          hasPrev: data['data']['has_prev'] ?? false,
        );
      }
    }
    throw Exception('Failed to load customers');
  }

  /// Fetch a single customer by ID.
  Future<Customer> getCustomerById(String id) async {
    final response = await _client.get('/customers/$id');

    if (response.statusCode == 200) {
      final data = response.data;
      if (data['success'] == true && data['data'] != null) {
        return Customer.fromJson(data['data']);
      }
    }
    throw Exception('Failed to load customer');
  }

  /// Create a new customer.
  Future<Customer> createCustomer(Map<String, dynamic> body) async {
    final response = await _client.post('/customers', data: body);

    if (response.statusCode == 200 || response.statusCode == 201) {
      final data = response.data;
      if (data['success'] == true && data['data'] != null) {
        return Customer.fromJson(data['data']);
      }
    }
    throw Exception('Failed to create customer');
  }

  /// Update an existing customer.
  Future<Customer> updateCustomer(String id, Map<String, dynamic> body) async {
    final response = await _client.patch('/customers/$id', data: body);

    if (response.statusCode == 200) {
      final data = response.data;
      if (data['success'] == true && data['data'] != null) {
        return Customer.fromJson(data['data']);
      }
    }
    throw Exception('Failed to update customer');
  }

  /// Delete a customer.
  Future<void> deleteCustomer(String id) async {
    final response = await _client.delete('/customers/$id');

    if (response.statusCode != 200) {
      throw Exception('Failed to delete customer');
    }
  }
}

File: apps/admin/repository/customerRepository.ts
/**
 * Customer Repository
 *
 * Data access layer for customer entities using Supabase.
 * Implements repository pattern for clean separation of concerns.
 *
 * @module repository/customerRepository
 */

import { BaseRepository, RepositoryResult } from './supabaseClient';
import { 
  Customer, 
  CreateCustomerDTO, 
  UpdateCustomerDTO, 
  CustomerSearchParams, 
  CustomerSearchResult,
} from '@/domain';

export class CustomerRepository extends BaseRepository {
  private readonly tableName = 'customers';

  /**
   * Find all customers with optional filtering and pagination
   */
  async findAll(params: CustomerSearchParams = {}): Promise<RepositoryResult<CustomerSearchResult>> {
    const {
      query,
      phone,
      sort_by = 'created_at',
      sort_order = 'desc',
      page = 1,
      limit = 20,
    } = params;

    const offset = (page - 1) * limit;

    // Build filters
    const filters: Record<string, any> = {};
    if (phone) filters.phone = phone;

    // Build the main query
    let selectQuery = this.buildQuery(this.tableName, {
      select: '*',
      filters,
      orderBy: { column: sort_by, ascending: sort_order === 'asc' },
      limit,
      offset,
      count: 'exact',
    });

    // Apply text search on name or email
    if (query) {
      selectQuery = (selectQuery as any).or(`name.ilike.%${query}%,email.ilike.%${query}%,phone.ilike.%${query}%`);
    }

    // Execute consolidated query (gets both data and total exact count in a single trip)
    const response = await selectQuery;
    const { data, count, error } = response as any;
    const result = this.handleResponse<Customer[]>({ data, error });

    if (!result.success || !result.data) {
      return {
        data: null,
        error: result.error,
        success: false,
      };
    }

    const total = count || 0;
    const totalPages = Math.ceil(total / limit);

    const searchResult: CustomerSearchResult = {
      customers: result.data,
      total,
      page,
      limit,
      total_pages: totalPages,
      has_next: page < totalPages,
      has_prev: page > 1,
    };

    return {
      data: searchResult,
      error: null,
      success: true,
    };
  }

  /**
   * Find customer by ID
   */
  async findById(id: string): Promise<RepositoryResult<Customer>> {
    return this.executeOperation<Customer>(async () => {
      const response = await this.client
        .from(this.tableName)
        .select('*')
        .eq('id', id)
        .single();
        
      return response;
    });
  }

  /**
   * Find customer by exact phone number
   */
  async findByPhone(phone: string): Promise<RepositoryResult<Customer>> {
    return this.executeOperation<Customer>(async () => {
      const response = await this.client
        .from(this.tableName)
        .select('*')
        .eq('phone', phone)
        .maybeSingle(); // Use maybeSingle to not error on 0 rows
        
      return response;
    });
  }

  /**
   * Create a new customer
   */
  async create(data: CreateCustomerDTO): Promise<RepositoryResult<Customer>> {
    return this.executeOperation<Customer>(async () => {
      const response = await this.client
        .from(this.tableName)
        .insert([data])
        .select()
        .single();
        
      return response;
    });
  }

  /**
   * Update an existing customer
   */
  async update(id: string, data: UpdateCustomerDTO): Promise<RepositoryResult<Customer>> {
    return this.executeOperation<Customer>(async () => {
      const response = await this.client
        .from(this.tableName)
        .update(data)
        .eq('id', id)
        .select()
        .single();
        
      return response;
    });
  }

  /**
   * Delete a customer
   */
  async delete(id: string): Promise<RepositoryResult<boolean>> {
    return this.executeOperation<boolean>(async () => {
      const { error } = await this.client
        .from(this.tableName)
        .delete()
        .eq('id', id);

      if (error) throw error;
      return { data: true, error: null };
    });
  }

  /**
   * Check if a customer has any existing orders
   * Used for delete safety check
   */
  async hasOrders(id: string): Promise<boolean> {
    const { count } = await this.client
      .from('orders')
      .select('*', { count: 'exact', head: true })
      .eq('customer_id', id);

    return (count || 0) > 0;
  }
}

// Export singleton instance
export const customerRepository = new CustomerRepository();
